code/modules/exonet/mainframe.dm

```diff
@@ -0,0 +1,162 @@
+/obj/machinery/computer/exonet/mainframe
+	name = "\improper EXONET Mainframe"
+	desc = "A very complex mainframe capable of storing massive amounts of data. Looks fragile."
+	ui_template = "exonet_mainframe.tmpl"
+	active_power_usage = 10 KILOWATTS
+	maximum_component_parts = list(
+		/obj/item/stock_parts = 10,
+		/obj/item/stock_parts/exonet_lock/buildable = 1,
+		/obj/item/stock_parts/computer/hard_drive = 6
+	)
+	var/list/initial_programs = list(
+		/datum/computer_file/program/email_client,
+		/datum/computer_file/program/wordprocessor,
+		/datum/computer_file/program/game,
+		/datum/computer_file/program/device_editor
+	)		// Optional variable for starting a mainframe with some programs in it.
+	var/setting_max_log_count = 100
+
+	var/max_capacity = 0
+	var/used_capacity = 0
+	var/list/stored_files = list()		// List of stored files on this mainframe. DO NOT MODIFY DIRECTLY!
+
+/obj/machinery/computer/exonet/mainframe/Initialize()
+	. = ..()
+	if(initial_programs)
+		for(var/initial_program in initial_programs)
+			var/datum/computer_file/program/prog = new initial_program
+			if(!prog || !istype(prog) || prog.filename == "UnknownProgram" || prog.filetype != "PRG")
+				continue
+			LAZYDISTINCTADD(stored_files, prog)
+		initial_programs = null
+	recalculate_size()
+
+/obj/machinery/computer/exonet/mainframe/proc/get_available_software()
+	var/list/programs = list()
+	for(var/datum/computer_file/program/prog in stored_files)
+		LAZYDISTINCTADD(programs, prog)
+	return programs
+
+/obj/machinery/computer/exonet/mainframe/proc/get_log_file()
+	var/datum/computer_file/data/logfile/file = find_file_by_name("network_log")
+	if(!file)
+		file = new()
+		file.filename = "network_log"
+		store_file(file)
+	return file
+
+// Use this proc to add file to the drive. Returns 1 on success and 0 on failure. Contains necessary sanity checks.
+/obj/machinery/computer/exonet/mainframe/proc/store_file(var/datum/computer_file/F)
```

Comment:
wouldn't whole thing be easier if it'd be using ntos for file stuff? Seems like it's pretty much reimplementing all the filessystem stuff from there.

Reply:
Actually is a copy of hard_drive. Reason that is, is because this is an equivalent LVM. There's a TODO for future to add in some RAID-type simulation and actually write down to the HDD level but the interface needed exposed on a machine level.